addons/udes_stock/tools/retry.py

```diff
@@ -0,0 +1,58 @@
+from functools import wraps
+import logging
+import time
+from psycopg2 import OperationalError, errorcodes
+
+_logger = logging.getLogger(__name__)
+
+PG_CONCURRENCY_ERRORS_TO_RETRY = (
+    errorcodes.LOCK_NOT_AVAILABLE,
+    errorcodes.SERIALIZATION_FAILURE,
+    errorcodes.DEADLOCK_DETECTED,
+)
+
+
+def retry_on_concurrency_error(max_retries=5):
+    """ Decorator to retry an operation when concurrency errors occur. """
+
+    def inner_function(func):
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            return _retry_on_concurrency_error(
+                self, lambda: func(self, *args, **kwargs), max_retries
+            )
+
+        return wrapper
+
+    return inner_function
+
+
+def _retry_on_concurrency_error(self, f, max_retries=5):
+    for tries in range(max_retries + 1):
+        try:
+            with self.env.cr.savepoint():
+                return f()
+        except OperationalError as e:
+            self.invalidate_cache()
+            if e.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
+                # don't do this
+                raise
+            if tries >= max_retries:
+                _logger.info(
+                    "%s, maximum number of tries reached"
+                    % errorcodes.lookup(e.pgcode)
+                )
+                raise
+
+            wait_time = 1
```

Comment:
Do we want `wait_time` as another function argumen as for `max_retries`? It'd give us more flexibility for longer running processes.

Reply:
Good idea, I'll add it

Reply:
Good point. A parameter looks reasonable to me.

In fact, wait time usually multiplies by 2 after every retry, so after each retry waits a bit longer. So it would be good to have another parameter like "delaying factor" so by default is 2, but in some cases we might want to keep it the same (so we can use 1 as factor).